src/Nethermind/Nethermind.Consensus.AuRa/Contracts/ReportingValidatorContract.cs

```diff
@@ -14,26 +14,84 @@
+        /// Called by the node of each honest validator after the specified validator misbehaved.
+        /// <seealso>
+        ///     <cref>https://openethereum.github.io/wiki/Validator-Set.html#reporting-contract</cref>
+        /// </seealso>
+        /// </summary>
+        /// <param name="maliciousMiningAddress">The mining address of the malicious validator.</param>
+        /// <param name="blockNumber">The block number where the misbehavior was observed.</param>
+        /// <returns>Transaction to be added to pool.</returns>
+        Transaction ReportBenign(Address maliciousMiningAddress, UInt256 blockNumber);
+    }
+
+    public sealed class ReportingValidatorContract : Contract, IReportingValidatorContract
+    {
+        private readonly ISigner _signer;
+        public Address NodeAddress => _signer.SigningAddress;
         
         public ReportingValidatorContract(
-            ITransactionProcessor transactionProcessor, 
             IAbiEncoder abiEncoder, 
-            Address contractAddress, 
-            IStateProvider stateProvider, 
-            IReadOnlyTransactionProcessorSource readOnlyReadOnlyTransactionProcessorSource) 
-            : base(transactionProcessor, abiEncoder, contractAddress, stateProvider, readOnlyReadOnlyTransactionProcessorSource)
+            Address contractAddress,
+            ISigner signer)
+            : base(abiEncoder, contractAddress)
         {
+            _signer = signer ?? throw new ArgumentNullException(nameof(signer));
         }
+
+        /// <summary>
+        /// Reports that the malicious validator misbehaved at the specified block.
+        /// Called by the node of each honest validator after the specified validator misbehaved.
+        /// <seealso>
+        ///     <cref>https://openethereum.github.io/wiki/Validator-Set.html#reporting-contract</cref>
+        /// </seealso>
+        /// </summary>
+        /// <param name="maliciousMiningAddress">The mining address of the malicious validator.</param>
+        /// <param name="blockNumber">The block number where the misbehavior was observed.</param>
+        /// <param name="proof">Proof of misbehavior.</param>
+        /// <returns>Transaction to be added to pool.</returns>
+        public Transaction ReportMalicious(Address maliciousMiningAddress, UInt256 blockNumber, byte[] proof) => GenerateTransaction<GeneratedTransaction>(nameof(ReportMalicious), NodeAddress, maliciousMiningAddress, blockNumber, proof);
+
+        /// <summary>
+        /// Reports that the benign validator misbehaved at the specified block.
+        /// Called by the node of each honest validator after the specified validator misbehaved.
+        /// <seealso>
+        ///     <cref>https://openethereum.github.io/wiki/Validator-Set.html#reporting-contract</cref>
+        /// </seealso>
+        /// </summary>
+        /// <param name="maliciousMiningAddress">The mining address of the malicious validator.</param>
```

Comment:
maliciousMinerAddress means that the Miner is malicious
maliciousMiningAddress means that the Address is malicious